docs/internal/abi.md

```diff
@@ -0,0 +1,437 @@
+[0][**][1][0101] USize (on a 64 bit platform)
+```
+
+This bit pattern is also arbitrary and can change in the actual implementation.
+
+Note that, in 64-bit architectures, pointers (and `F64`) have the same width on both native and NaN-boxed representations.
+All other types increase the width to at least 64-bit, even if they were `U8` to begin with.
+
+Also note that the pattern above is an extra byte, so all values will have the whole 64/128 bits of the payload to be encoded.
+
+## Arrays
+
+Arrays are structures that contain a dynamically sized list of elements of a single type.
+The memory layout of the array is guaranteed to be consecutive (ex. `addr(array[N+1]) == addr(array[N] + sizeof(type)`).
+
+Example:
+```ts
+// No surprises here
+Array[U16] -> [ i16, i16, i16, ...]
+Array[A] -> [ ptr, ptr, ptr, ... ]
+
+// Naive Bool array
+Array[Bool] -> [ i8, i8, i8, ... ]
+
+// Compact Bool array
+Array[Bool] -> [ { i1, i1, i1, ... }, { i1, ... } ... ]
+```
+
+That type, however, can be a union type, so it needs descriptors.
+The naive layout for an array of unions is to have an array of packed descriptors.
+
+Example:
+```ts
+// NaN-boxed types
+Array[U16 | I8] -> [ i64, i64, i64, ... ]
+Array[U32 | F64] -> [ i64, i64, i64, ... ]
+Array[A | F16] -> [ i64, i64, i64, ... ]
+
+// wide-packed types
+Array[A | U64] -> [ { i8, i64 }, { i8, i64 }, ... ]
+Array[F64 | U128] -> [ { i8, i128 }, { i8, i128 }, ... ]
+```
+
+NaN-boxed types have the problem that it increases the size of the object at least to 64-bits (128-bit on CHERI), even when neither of the types are 64-bit wide.
+This may not be a problem for individual objects, but an array with multiple objects will multiply the problem.
+
+Example:
+```ts
+// Single-types
+Array[I8] -> [ i8, i8, i8, ... ]
+Array[U8] -> [ i8, i8, i8, ... ]
+
+// NaN-boxed types
+Array[U8 | I8] -> [ i64, i64, i64, ... ]
+```
+
+That, however, is a small problem when compared to the wide packing.
+Not only the wide-packed array increases the payload, but it also introduces the descriptor which is not the same size as the payload.
+This creates misalignment issues (on platforms where misaligned reads are penalised or forbidden, this is really bad).
+
+An optimisation under discussion is to pack the descriptors into a payload-sized element.
+With a 64-bit payload, the first 8 elements would follow the first 64-bit bucket with their 8 descriptors packed.
+The following 8 elements would need a new descriptor bucket, placed after the 8th element, and so on.
+
+Example:
+```ts
+Array[F64 | U128] -> [ { i8, i64 }, { i8, i64 }, ... ]
+
+// to
+
+type Desc = { i8, i8, i8, i8, i8, i8, i8, i8 };
+Array[F64 | U128] -> [ Desc, i64, i64, i64, i64, i64, i64, i64, i64, Desc, i64, ...  ]
+```
+
+To access the 7-th element, the unpacking does:
+```ts
+// Takes the 7th descriptor (in pseudo tuple notation starting at _0)
+descrip = Array[0]._6;
+
+// Takes the 7th element (at 8th position starting from 0)
+payload = Array[6+1];
+```
```

Comment:
```suggestion

To facilitate vectorisation, it might be more efficient to store parallel arrays of values and arrays of discriminators.
It is possible to optimise this further for cases where the total number of machine-word types in the array is limited.
For example `Array[Foo | U64]` requires only one bit of discriminator state to identify the type.
This could be compressed by providing a two-element map (16 bits of state in the array object) that indexes from a dense type representation to a the wide encoding.
In this form, an `Array[Foo | U64]` would be encoded as a structure of the form:

```llvm
array = {{ ptr, ptr }, i1*, ptr*, i64, i8[2] }
```

The first two pointers are the object header, the next is a pointer to an array of one-bit discriminators for the different types, then a pointer to an array of payloads, followed by the length of the array and a map from the `i1` discriminators to the wide discriminators.
Accessing array element 4 would require the following steps:

 - Load the element at offset 3 in the payload array.
 - Load the bit at offset 3 in the discriminator array.
 - Load the real discriminator from the start of the discriminator map + the value loaded from the discriminator array.

In a loop, all of these operations have a trivial mapping from the loop induction variable and so are amenable to vectorisation.
```

Reply:
Github doesn't do nice nesting of code things, annoyingly, so the above suggestion is mangled.

Reply:
That's a good idea, too, though I think the example is too confusing for this document. It would require a longer explanation. I think we can just leave the description and let the actual representation to be discussed later.

wrt formatting, I'll copy locally, check formatting and commit in separate.